### PR TITLE
[AIRFLOW-2303] S3 List Operator

### PR DESCRIPTION
**Pull request by wileeam**
Make sure you have checked _all_ steps below.

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title: "\[AIRFLOW-2303\] S3 List Operator"
    - https://issues.apache.org/jira/browse/AIRFLOW-2303

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
- Lists the contents of an S3 bucket given a prefix and a delimiter (both optional) and returns the contents as a list.

### Tests
- [x] My PR adds the following unit tests:
- `tests/contrib/operators/test_s3_list_operator.py`
- Tested also in house (directly with an S3 bucket)

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: docs/integration.rst [hunk 1]
```diff
@@ -196,10 +196,18 @@ EmrHook
...
 
+- :ref:`S3ListOperator` : Lists the files matching a key prefix from a S3 location.
 - :ref:`S3FileTransformOperator` : Copies data from a source S3 location to a temporary location on the local filesystem.
 - :ref:`S3ToHiveTransfer` : Moves data from S3 to Hive. The operator downloads a file from S3, stores the file locally before loading it into a Hive table.
 - :ref:`S3Hook` : Interact with AWS S3.
 
+.. _S3ListOperator:
+
+S3ListOperator
+"""""""""""""""
```
Comment (Fokko) [inline docs/integration.rst]: Please remove `"` to match the length

File: docs/code.rst [hunk 2]
```diff
@@ -164,6 +164,7 @@ Operators
 .. autoclass:: airflow.contrib.operators.pubsub_operator.PubSubSubscriptionDeleteOperator
 .. autoclass:: airflow.contrib.operators.pubsub_operator.PubSubPublishOperator
 .. autoclass:: airflow.contrib.operators.qubole_operator.QuboleOperator
+.. autoclass:: airflow.operators.s3_list_operator.S3ListOperator
```
Comment (Fokko) [inline docs/code.rst]: You're missing `contrib` here

Comment (wileeam) [reply docs/code.rst]: Fixed!

File: docs/integration.rst (hunk 1 above)
Comment (wileeam) [reply docs/integration.rst]: Fixed!

**Comment by Fokko** [comment]
Can you rebase onto master?

**Comment by wileeam** [comment]
Done @Fokko 
Tests fired up again.

**Review by Fokko — APPROVED** [review]
LGTM, thanks @wileeam

**Comment by Fokko** [comment]
@wileeam I've reworked the commit message a bit. For the future, please align with the git message conventions in the PR-template.

**Comment by wileeam** [comment]
Thanks @Fokko!
And understood... I missed that this time. I will abide by the conventions in the upcoming PR based on this one, now that is merged :)

**Pull request closed without merging**